File: 0093-restore-ip-addresses/python/Solution.py
class Solution:
    def __init__(self):
        self.result = []

    def restoreIpAddresses(self, s: str) -> List[str]:
        self.backtrack(s, 0,0, [])
        return self.result

    def backtrack(self, s, k, step, path):
        if step == 4 and k == len(s):
            sb = [str(item) for item in path]
            self.result.append(".".join(sb))
            return
        if step > 4:
            return
        if k == len(s):
            return
        val = 0
        if k < len(s):
            val = val * 10 + int(s[k])
            path.append(val)
            self.backtrack(s, k+1, step+1, path)
            path.pop()
        if s[k] == "0":
            return
        if k + 1 < len(s):
            val = val * 10 + int(s[k+1])
            path.append(val)
            self.backtrack(s, k+2, step+1, path)
            path.pop()
        if k + 2 < len(s):
            val = val * 10 + int(s[k+2])
            if val <= 255:
                path.append(val)
                self.backtrack(s, k+3, step+1, path)
                path.pop()

File: 0093-restore-ip-addresses/javascript/Solution.js
/**
 * @param {string} s
 * @return {string[]}
 */
var restoreIpAddresses = function(s) {
    let result = []
    let path = []
    let backtrack = (k, step) => {
        if (k == s.length && step == 4) {
            result.push(path.join("."))
            return
        }
        if (step > 4) {
            return
        }
        if (k == s.length) {
            return
        }
        let val = 0
        if (k < s.length) {
            val = val * 10 + (s[k].charCodeAt(0)-'0'.charCodeAt(0))
            path.push(val)
            backtrack(k + 1, step + 1)
            path.pop()
        }
        if (s[k] == '0') {
            return
        }
        if (k+1 < s.length) {
            val = val * 10 + (s[k+1].charCodeAt(0) - '0'.charCodeAt(0))
            path.push(val)
            backtrack(k+2, step+1)
            path.pop()
        }
        if (k+2 < s.length) {
            val = val * 10 + (s[k+2].charCodeAt(0) - '0'.charCodeAt(0))
            if (val <= 255) {
                path.push(val)
                backtrack(k+3, step+1)
                path.pop()
            }
        }
    }
    backtrack(0, 0)
    return result
};